drivers/hpet.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright © 2020 Amazon.com, Inc. or its affiliates.
+ * All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <drivers/hpet.h>
+#include <ioapic.h>
+
+bool init_hpet(uint8_t dst_cpus) {
+    printk("Initializing HPET\n");
+    acpi_hpet_t *hpet = (acpi_hpet_t *) acpi_find_table(HPET_SIGNATURE);
+    if (hpet) {
+        mfn_t hpet_base_mfn = paddr_to_mfn(hpet->address.address);
+        vmap((uint64_t *) hpet->address.address, hpet_base_mfn, PAGE_ORDER_4K, L1_PROT);
+        volatile acpi_hpet_timer_t *config =
+            (acpi_hpet_timer_t *) (hpet->address.address +
+                                   ACPI_HPET_OFFSET_TIMER_0_CONFIG_CAP_REG);
+        volatile acpi_hpet_general_t *general =
+            (acpi_hpet_general_t *) (hpet->address.address +
+                                     ACPI_HPET_OFFSET_GENERAL_CAP_REG);
+        volatile uint64_t *main_counter =
```

Comment:
nit: please move declaration of these variables to the function prolog, it will make this line shorter